x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/GroupByOptimizer.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.transform.transforms.pivot;
+
+import org.elasticsearch.xpack.core.transform.transforms.pivot.SingleGroupSource;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+public final class GroupByOptimizer {
+
+    private static class UnmodifiableEntryWithPriority implements Entry<String, SingleGroupSource> {
+
+        private final Entry<String, SingleGroupSource> entry;
+        private final int priority;
+
+        UnmodifiableEntryWithPriority(Entry<String, SingleGroupSource> entry, int priority) {
+            this.entry = entry;
+            this.priority = priority;
+        }
+
+        public int compareTo(UnmodifiableEntryWithPriority o) {
+            return Integer.compare(priority, o.priority);
+        }
+
+        @Override
+        public String getKey() {
+            return entry.getKey();
+        }
+
+        @Override
+        public SingleGroupSource getValue() {
+            return entry.getValue();
+        }
+
+        @Override
+        public SingleGroupSource setValue(SingleGroupSource value) {
+            // not modifiable
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private GroupByOptimizer() {}
+
+    /**
+     * Returns an ordered collection of group by fields in order to get better performance.
+     *
+     * The decision is based on the type and whether the input field is a indexed/runtime/script field
+     *
+     * TODO: take index sorting into account
+     *
+     * @param groups group by as defined by the user
+     * @param runtimeFields set of strings which fields are runtime fields
+     * @return collection in order of priority
+     */
+    static Collection<Entry<String, SingleGroupSource>> reorderGroups(Map<String, SingleGroupSource> groups, Set<String> runtimeFields) {
+        if (groups.size() == 1) {
+            return groups.entrySet();
+        }
+
+        // Arrays.sort provides stable sort (to respect the input order if priorities match), Collections.sort not
+        UnmodifiableEntryWithPriority[] prioritizedGroups = new UnmodifiableEntryWithPriority[groups.size()];
```

Comment:
`UnmodifiableEntryWithPriority` could be replaced with `Tuple<Entry<String, SingleGroupSource>, Integer>`

Then you could do `Comparitor.comparing(Tuple::v2)`. The static class seems like a ton of unnecessary code. But this is a minor quibble.

Reply:
I had something like that in one of the 1st versions (before realizing the problem with sort), however generic types and generic arrays don't work together (I could suppress it, but don't like to).

Reply:
> however generic types and generic arrays don't work together 

Very true.